### PR TITLE
Add `attributes` free-form map field for tool-specific fields in a number of places

### PR DESCRIPTION
**Pull request by davidfestal**
### What does this PR do?

This PR adds the `attributes` map, to enable adding tool-specific fields, at the root of the following objects:
components, projects, starterProjects and commands.

The `attributes` map is not only a map of strings, but also allows adding any json-schema values in map values (YAML values in yaml documents or Json values in JSON documents). here is an example:

```yaml
components:
  - name: a-container
    attributes:
      tool1-string-attribute: a-string
      tool2-object-attribute:
        free-form-string-field: "value" 
        free-form-boolean-field: true 
      tool3-array-attribute:
        - value1 
        - value2 
```

This PR also updates the `attributes` already present in the container `endpoints` definition to allow objects in map values. This is **not a breaking change** though since the current proposal also covers the "map of strings" case.

In go the attribute values would be retrieved through one of the `Attributes` methods, some of which are demonstrated in the following snippet:

```go
component : ...
attributes := component.Attributes

var err error

withoutErrorChecking := attributes.GetString("tool1-string-attribute")

withErrorChecking := attributes.GetString("tool1-string-attribute", &err)

myStruct := struct {
  StringField string `json:"free-form-string-field"`
  BooleanField string `json:"free-form-boolean-field"`
}

err = attributes.GetInto("tool2-object-attribute", &myStruct)

var arrayValues []interface{} = attributes.Get("tool3-array-attribute")
```
 
### What issues does this PR fix or reference?

This fixes issue https://github.com/devfile/api/issues/169

### Is your PR tested? Consider putting some instruction how to test your changes

Yes, it is.

### CODE REVIEW

File: pkg/apis/workspaces/v1alpha2/attributes.go
Comment (sleshchenko) [inline pkg/apis/workspaces/v1alpha2/attributes.go:9]: from the model point of view, it's so useful to have any JSON in the field, but from client(odo, devworkspace) perspective of view, it means that schema does not validate attributes format and every field access requires checking that attributes have the right format, if I did not miss something. So, we need some library code that allows easily to get attribute value as string, int, interface or error if the field has the wrong format.
Am I missing something and it's not going to be an issue for clients to read the attributes?

Comment (davidfestal) [reply pkg/apis/workspaces/v1alpha2/attributes.go:9]: You're right, I'm looking into it.

Comment (amisevsk) [reply pkg/apis/workspaces/v1alpha2/attributes.go:9]: To add here, I don't know how elegant or useful embedding full json as an attribute will ultimately be. Even with `map[string]string` you _could_ ostensibly use json, but the main problem is that it's hard to write. I think I'd rather see something like
```yaml
attributes:
  controller.devfile.io/plugin/name: "vscode-java"
  controller.devfile.io/plugin/vscode-extensions/java: "https://my-url.com/java.vsix"
  controller.devfile.io/plugin/vscode-extensions/java-debug: "https://my-url.com/java-debug.vsix"
```
than have the above collapsed into a json blob
```yaml
attributes:
  controller.devfile.io/plugin: '{"name": "vscode-java", "plugins": ["java.vsix", "java-debug.vsix"]}'
```
The latter could get quite complicated as there are longer entries. It also forces attributes to be complex, as we couldn't use something like `my-attribute: my-value`.

**Review by amisevsk — COMMENTED** [review]
I have no suggestions apart from the point brought up by Serhii

File: pkg/apis/workspaces/v1alpha2/attributes.go
Comment (davidfestal) [reply pkg/apis/workspaces/v1alpha2/attributes.go:9]: > than have the above collapsed into a json blob

@amisevsk we just don't need to have json collasped into a JSON blob.
This is a json-like arbitrary value, directly included inside the JSON or YAML document.
So in a YAML document it will a yaml fragment, and in a JSON document it will be a JSON document.

Your example would be:

```yaml
    attributes:
      name: vscode-java
      plugins:
        - java.vsix
        - java-debug.vsix
```

That's the whole point of the change.

Comment (davidfestal) [reply pkg/apis/workspaces/v1alpha2/attributes.go:9]: And with the helpers I'm currently adding to the PR, when getting the GO structs, you would extract it like that:

```go
var Attributes attributes = ...
myName, isAString := attributes.Get("name").Interface().(string)

// And for plugins you might use want to use the `DecodeInto` method, to decode the arbitrary JSON-like structure into a given, well-known GO type. 

plugins := &[]string{}
err := attributes.Get("plugins").DecodeInto(plugins)

// Of course you also have a method to know if a given key is available in the Attributes:

isThere := attributes.Exists("aMissingKey")
```

Comment (amisevsk) [reply pkg/apis/workspaces/v1alpha2/attributes.go:9]: This makes sense to me and is a lot cleaner than my suggestion :)

**Comment by davidfestal** [comment]
I simplified the `attributes` api helpers and added many tests.

@amisevsk @sleshchenko on this personnal branch https://github.com/amisevsk/devworkspace-operator/compare/use-v1alpha2...davidfestal:use-v1alpha2-and-new-attributes you will find examples of use in the changes I made on @amisevsk `use-v1alpha2` to accomodate the new attributes in the `endpoints` case.

Still need to document public methods in the new `attributes` package though.

File: pkg/apis/workspaces/v1alpha2/commands.go
Comment (amisevsk) [inline pkg/apis/workspaces/v1alpha2/commands.go:65]: Is this formatted? I don't know if field type should align with struct tags below

File: pkg/attributes/attributes.go
Comment (amisevsk) [inline pkg/attributes/attributes.go:20]: I'd suggest `DecodeKeyInto` as a function name here, or something similar. This name might be too similar to `DecodeInto` below.

Also I think `DecodeInto` and `GetDecodedInto` should be adjacent in this file.

File: pkg/attributes/attributes.go
Comment (amisevsk) [inline pkg/attributes/attributes.go:43]: I think we should return the error instead of passing in a struct optionally. This approach makes it very easy to hide ignoring a potential error.

File: pkg/apis/workspaces/v1alpha2/attributes.go
Comment (maysunfaisal) [reply pkg/apis/workspaces/v1alpha2/attributes.go]: I am wondering if it makes sense to have these helper functions in devfile/api or devfile/library

**Review by amisevsk — APPROVED** [review]
Generally LGTM

File: pkg/apis/workspaces/v1alpha2/attributes.go
Comment (davidfestal) [reply pkg/apis/workspaces/v1alpha2/attributes.go]: @maysunfaisal In fact it is there, because my plan was to also use it in the `DevfileMetadata` struct, which is also in the `devfile/api` repo, but in the non-K8S-versioned `devfile` package.

File: pkg/devfile/header.go
Comment (maysunfaisal) [inline pkg/devfile/header.go:30]: Should it be `json:"attributes,omitempty"`

Comment (davidfestal) [reply pkg/devfile/header.go]: I was trying to have the free-form attributes at the same level as the main metadata, but this is obviously not really possible.
I finally changed it according to your suggestion in my last commit:  https://github.com/devfile/api/pull/214/commits/1a8f497e4106b5c5b23b61d020b0d9df4799e455

File: pkg/apis/workspaces/v1alpha2/commands.go
Comment (davidfestal) [reply pkg/apis/workspaces/v1alpha2/commands.go:65]: It is yes. That's probably a bug of the formatter itself.

File: pkg/attributes/attributes.go
Comment (davidfestal) [reply pkg/attributes/attributes.go:57]: The whole point of this approach is to avoid requiring a multiple value return, in order to be able to chain calls when we know there is no reason for an error.

Comment (davidfestal) [reply pkg/attributes/attributes.go:57]: But here I can change it. I'd be more reluctant to change it for the `GetString()`... methods, since it's quite useful to be able to chain them or directly use the result in an expression.

**Review by amisevsk — APPROVED** [review]
LGTM after updates.

File: pkg/attributes/attributes.go
Comment (davidfestal) [reply pkg/attributes/attributes.go:181]: I thought about gathering errors, but didn't want to add a dependency to a non-standard error package to tackle error gathering. Didn't put much thought in it though.
So for now only the last error is returned.

File: pkg/attributes/attributes.go
Comment (sleshchenko) [inline pkg/attributes/attributes.go:94]: IDE suggests that `Condition is always 'true' because 'errorHolder' is always not 'nil'` here and in other places in this file.

File: pkg/attributes/attributes.go
Comment (sleshchenko) [reply pkg/attributes/attributes.go:181]: OK with that, but maybe it worths a few words in the documentation.

Comment (davidfestal) [reply pkg/attributes/attributes.go:181]: You're right. Let me do it.

File: pkg/attributes/attributes.go
Comment (sleshchenko) [reply pkg/attributes/attributes.go]: > when we know there is no reason for an error

Can we identify when we know when there is a reason for the error and when there is no?
Like in the shared example https://github.com/amisevsk/devworkspace-operator/compare/use-v1alpha2...davidfestal:use-v1alpha2-and-new-attributes#diff-cee3fd6d4114264b1f235c57bd27c71a9c40cd809efa214a395f5e4cfee8a92fR80
is it safe to propagate an empty string in case the user misconfigured protocol or path as an object?
It shows that such an aproach forces client not to check the erorr. As an alternative, we could have two separate methods(for each type) one would return the error, another will ignore the error but would be named in the way client understand error is suppressed. Like GetStringOrDefault(), default depends on type: string - "", int - 0, interface - {}.

File: pkg/attributes/attributes.go
Comment (davidfestal) [reply pkg/attributes/attributes.go:190]: Add comments whenever relevant in commit https://github.com/devfile/api/pull/214/commits/cf8f396b3d6352db2280c80030c5d3e6a368eeb3

File: pkg/attributes/attributes.go
Comment (davidfestal) [reply pkg/attributes/attributes.go]: Good catch ! Thanks !

Comment (davidfestal) [reply pkg/attributes/attributes.go]: Fixed in commit https://github.com/devfile/api/pull/214/commits/ed4b1865c5d7ee354d764eb777f116dba96af9d4

File: pkg/attributes/attributes.go
Comment (sleshchenko) [reply pkg/attributes/attributes.go]: More relevant case, where some boolean value is mistyped by user, like public: "true", instead of public: true. Then user will get unexpected behavior instead of error:
```go
func TestBooleanAsString(t *testing.T) {
	stringMap := Attributes{}.FromStringMap(map[string]string{
		"public": "true",
	})
	publicAttr := stringMap.GetBoolean("public")
	assert.Equal(t, true, publicAttr)
}
```

**Comment by davidfestal** [comment]
@maysunfaisal @elsony @l0rd @yangcao77 

Do you want to review this PR after the very last changes ?

**Review by sleshchenko — APPROVED** [review]
Good job.

I don't have the ideal solution to propose with error handling, so approving.

About string/boolean... Maybe to avoid unexpected behavior, we need to implement autoconversion between primitive types, because if the format allows and user defines like `true` instead of `"true"`, or `"123"` instead of `123`, there is no a real reason why we should fail it with a validation error or start with default values.
Or we just need to declare that clients need to validate attributes they expect before using, for devworkspace controller it may be done with webhooks and then it should be pretty safe to ignore error on reconciling.

**Comment by davidfestal** [comment]
> About string/boolean... Maybe to avoid unexpected behavior, we need to implement autoconversion between primitive types, because if the format allows and user defines like `true` instead of `"true"`, or `"123"` instead of `123`, there is no a real reason why we should fail it with a validation error.

I've implemented that in commit https://github.com/devfile/api/pull/214/commits/5640a2403c18ceaf91749b0bcb47922f948718e9
and added relevant tests.

@amisevsk @sleshchenko Do you think you could add a last review step after this commit ?

**Comment by maysunfaisal** [comment]
@davidfestal looks good 👍

**Review by maysunfaisal — APPROVED** [review]

**Review by yangcao77 — APPROVED** [review]
thanks David for this PR, good changes! lgtm

File: pkg/attributes/attributes.go
Comment (amisevsk) [inline pkg/attributes/attributes.go:38]: Couldn't we derive `resultType` from `zeroValue`? Would it be possible to implement this function as
```go
func (attributes Attributes) getPrimitive(key string, into interface{}, convert convertPrimitiveFunc, errorHolder ...*error) {
//....
}
```
so that it could e.g. be called
```go
var value string
attributes.getPrimitive(myKey, &value, ...)
```
switching based on a raw string describing the type feels very similar to just switching on a `reflect.Type`

File: pkg/attributes/attributes.go
Comment (amisevsk) [inline pkg/attributes/attributes.go:72]: I'm not clear on what we're doing here -- why are we passing in a variadic error if we only ever set the first element?

**Review by amisevsk — APPROVED** [review]
A couple of questions but otherwise LGTM.

File: pkg/attributes/attributes.go
Comment (sleshchenko) [reply pkg/attributes/attributes.go:72]: As far as I understand how optional parameters could be achieved in go. So, we could call specify `GetString()` instead of `GetString(nil)`. But it allows us to call `GetString(err, err, err, err)` and the only first is used.

I'm OK with both cases with their pros:
- `GetString()`: which is kind of syntax sugar;
- `GetString(nil)`: is a bit more explicit and suggests that error is ignored - in case you know parameter type;

File: pkg/attributes/attributes.go
Comment (sleshchenko) [inline pkg/attributes/attributes.go:141]:
```suggestion
// - strings "0, "f", "F", "FALSE", "false", "False" will be converted to a `false` boolean
```

File: pkg/attributes/attributes.go
Comment (davidfestal) [reply pkg/attributes/attributes.go:38]: well, I somehow tried, but since I didn't want to use reflection, and wanted to keep it simple (it's just an internal private helper used in 3 functions), I chose this design.

File: pkg/attributes/attributes.go
Comment (davidfestal) [reply pkg/attributes/attributes.go:72]: You're right @sleshchenko. I just used variadic parameter to provide an optional parameter.
However in this form this API also allows, in the future, passing an array of errors and possibly getting as many errors as the provide variadic list allows, instead of only returning the last error encountered.

**Review by sleshchenko — APPROVED** [review]

File: pkg/attributes/attributes.go
Comment (sleshchenko) [reply pkg/attributes/attributes.go:72]: >  passing an array of errors and possibly getting as many errors as the provide variadic list allows, instead of only returning the last error encountered.

Has client initialize errors instances? I mean the client is not able to consume the items provided in the array, is it?
If not, then how you see it in the future? Like client initializes 100 errors and some of them are modified?

Comment (davidfestal) [reply pkg/attributes/attributes.go:72]: @sleshchenko I'm OK to change it to `GetString(key string, errorHolder *error)`,
and enforce the client to explicitly pass a `nil` error pointer in the second argument to clearly state it ignores the error.

@maysunfaisal @yangcao77 @amisevsk any preference between the use of an optional parameter through the use of variadics or a single error pointer ?

File: pkg/attributes/attributes.go
Comment (amisevsk) [reply pkg/attributes/attributes.go:38]: Fair enough.

File: pkg/attributes/attributes.go
Comment (amisevsk) [reply pkg/attributes/attributes.go:72]: I think an optional parameter would be nice to have, but also allows us to implicitly skip error checking with no way to easily notice it in code. Without knowing the API we're using in detail, it might be hard to notice e.g. 
```go
  myString := attributes.GetString("myKey")
```
is discarding an error silently.

We're kind of departing general Go paradigms for error handling (which I'm not a fan of anyways), so it's hard to reason about what makes sense here. I'm generally fine with whatever we choose, but it might be confusing for (eventual) newcomers.

Comment (davidfestal) [reply pkg/attributes/attributes.go:72]: So maybe @sleshchenko proposal of using `GetString("aString", nil)` is the best here ?This would still allow using `GetString` in expressions (by avoiding multiple value returns) while avoiding silent error discarding ?

**Comment by davidfestal** [comment]
As deiscussed today, we're waiting for you GO to merge this @elsony, based of the devfile parser library readiness.

**Comment by elsony** [comment]
@davidfestal I have checked with the team. It turns out that the parser code is currently pulling in a specific tag of the schema so merging this change is not going to break the parser code. I also checked with @mmulholl and he'll adjust his test PR to accommodate that. Given that his test code has not been merged yet, so it is not going to be a problem. So feel free to merge this PR whenever you think is ready.

File: pkg/attributes/attributes.go
Comment (davidfestal) [reply pkg/attributes/attributes.go]: Done in commit https://github.com/devfile/api/pull/214/commits/2c5bf7950b093c7edbc8c77b3a6d2463e178ddc5

**Review by l0rd — APPROVED** [review]

**Pull request merged**
